apps/src/util/testJavabuilderWebsocketConnection.js

```diff
@@ -0,0 +1,51 @@
+import $ from 'jquery';
+
+// helper to allow behind the scenes test of javabuilder connection
+// on load of high traffic pages. particularly interested
+// in collecting data on whether schools often block websockets.
+export default function testJavabuilderWebsocketConnection() {
+  try {
+    logToFirehose('started');
+
+    const socket = new WebSocket(
+      'wss://javabuilderbeta.code.org?Authorization=connectivityTest'
+    );
+
+    socket.onopen = function(e) {
+      socket.send('connectivityTest');
+    };
+
+    socket.onmessage = function(message) {
+      if (message.data === 'success') {
+        logToFirehose('success');
+      } else {
+        logToFirehose('unexpected message response');
+      }
+      socket.close();
+    };
+
+    socket.onerror = function(error) {
+      logToFirehose('websocket error', error.toString());
+      socket.close();
+    };
+  } catch (error) {
+    logToFirehose('other error', error.toString());
+  }
+}
+
+// We log via our own servers to avoid
+// schools potentially blocking API calls to third parties,
+// which might affect our results if we were to log to firehose directly.
+const logToFirehose = (event, error) => {
+  const payload = {event};
+  if (error) {
+    payload.error = error;
+  }
+
+  return $.ajax({
```

Comment:
Sounds good -- we (CSA) just wrote a new [data API using jquery](https://github.com/code-dot-org/code-dot-org/blob/staging-next/apps/src/templates/codeReviewGroups/CodeReviewGroupsDataApi.js), so the preference is fetch going forward? FYI @code-dot-org/csa

Reply:
Good to know - @bencodeorg I think using fetch may also help us with the confusion we were running into between `.then()` and `.done()/.fail()`

Reply:
yep, the preference is to use fetch. @maureensturgeon ran into a [sneaky browser-specific issue](https://codedotorg.slack.com/archives/C0T0PNR0D/p1637774540054700) that she diagnosed/fixed, so calling that out so you don't run into the same

Reply:
mmm interesting, I had to explicitly specify x-csrf-token with fetch, which you mentioned in that thread ^^^. Had to move things around to make that work, take another look when you have a moment! I'm trying to understand why x-csrf-token was included in our jquery ajax calls, do we have a wrapper that adds it, or? I couldn't find anything.

Reply:
i'm not sure! i don't know of any code that automatically adds the CSRF token header to $.ajax calls

Reply:
I don't know much about CRSF, but as far as I know, it's up to us to decide whether we use a CRSF token, and if so, how we get it from the server to the client code.  Rails provides one and we can embed it in a rendered page how we like.  I thought it was more common to include it as a [hidden field](https://github.com/code-dot-org/code-dot-org/blob/cc99e4a86f5f2ebaa443339f3c93244f63a50b35/dashboard/app/views/layouts/_terms_interstitial.html.haml#L36) in a specific form, but it makes sense that we could also bake it into a meta tag if we need to grab it with some of our own JS.  We've sometimes built server APIs that [don't check it](https://github.com/code-dot-org/code-dot-org/search?q=skip_before_action++verify_authenticity_token), but it's overall a good thing to include where feasible.

Worth noting that we don't want to rely on it when posting from a page that might be cached, because we need a fresh token for each transmission to the server.  I'm also curious whether we're intending to use one token for multiple posts from one page... the token might no longer work after the first post?

Reply:
> I'm also curious whether we're intending to use one token for multiple posts from one page... the token might no longer work after the first post?

rails embeds the token in the markup that comes from the server, so the same token should work until you reload the page. we use this strategy (get ahold of the current token, store it in state, then use it for POST/DELETE requests) often throughout the frontend

Reply:
I was testing this on the applab side on a standalone project, which (obviously) is not cached, so I was able to access the csrf token via the `meta` tag. Testing on applab in the curriculum I realized this does not work, so I added a new route to fetch a csrf token before trying to log anything.

Reply:
i don't think we need an extra route/HTTP request -- script_levels have access to the CSRF token through `view_options`:

https://github.com/code-dot-org/code-dot-org/blob/9c5b3496acce81e1bdebd8481662fa0fee4c5000/dashboard/app/controllers/script_levels_controller.rb#L535

i added this to do something very similar, so linking that PR in case it's helpful: https://github.com/code-dot-org/code-dot-org/pull/38182

Reply:
if using that option is more pain than it's worth, i think it's fine to go with a separate route since this code is temporary. however, if it weren't temporary, we would want that route to be in a generic controller (rather than have the endpoint be javalab-specific)

Reply:
Ah that is good to know -- I'm inclined to leave as-is since this is temporary. I've seen csrf token fetching done in a lot of different ways, do you think we could incorporate a standard practice into the sort of "fetch post wrapper" that Maureen described on the Slack thread you linked?

Reply:
> yep, the preference is to use fetch. @maureensturgeon ran into a [sneaky browser-specific issue](https://codedotorg.slack.com/archives/C0T0PNR0D/p1637774540054700) that she diagnosed/fixed, so calling that out so you don't run into the same

Popping in to ask if this preference lives anywhere we can reference? Like maybe in a style guide or a "Front-End at Code.org" primer of some kind? I know we haven't standardized completely on a documentation plan, but this seems like a great nugget to include. Thanks!

Reply:
we've never really had a place to document information like this, but maybe [STYLEGUIDE.md](https://github.com/code-dot-org/code-dot-org/blob/staging/STYLEGUIDE.md) would be most appropriate for now?

Reply:
missed this in my last comment:

> do you think we could incorporate a standard practice into the sort of "fetch post wrapper" that Maureen described on the Slack thread you linked?

yes, i think that would be very helpful!